bctl/agent/plugin/shell/shell.go

```
@@ -0,0 +1,308 @@
+			rerr := fmt.Errorf("Malformed shell input payload %v", actionPayload)
+			k.logger.Error(rerr)
+			return action, []byte{}, rerr
+		}
+
+		if err := k.shellInput(shellInput); err != nil {
+			rerr := fmt.Errorf("Write to stdin failed %v", err)
+			k.logger.Error(rerr)
+			return action, []byte{}, rerr
+		}
+	case bzshell.ShellResize:
+		var shellResize bzshell.ShellResizeMessage
+
+		if err := json.Unmarshal(actionPayloadSafe, &shellResize); err != nil {
+			rerr := fmt.Errorf("Malformed shell resize payload %v", actionPayload)
+			k.logger.Error(rerr)
+			return action, []byte{}, rerr
+		}
+
+		if err := k.setSize(shellResize.Cols, shellResize.Rows); err != nil {
+			rerr := fmt.Errorf("Shell resize failed %v", err)
+			k.logger.Error(rerr)
+			return action, []byte{}, rerr
+		}
+	}
+
+	return action, []byte{}, nil
+}
+
+func (k *ShellPlugin) Ready() bool {
+	return !(k.stdin == nil || k.stdout == nil)
+}
+
+var startPty = func(
+	logger *logger.Logger,
+	runAsUser string,
+	commands string,
+	plugin *ShellPlugin) (err error) {
+
+	return StartPty(logger, runAsUser, commands, plugin)
+}
+
+func (k *ShellPlugin) open() error {
+	commands := ""
+
+	// Catch signals and send a signal to the "sigs" chan if it triggers
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM, syscall.SIGQUIT)
+
+	go func() {
+		sig := <-sigs
+		k.logger.Infof("caught signal to terminate: %v", sig)
+
+		if k.execCmd != nil {
+			if err := k.execCmd.Kill(); err != nil {
+				k.logger.Errorf("unable to terminate pty: %s", err)
+			}
+		}
+	}()
+
+	// Start pseudo terminal
+	if err := startPty(k.logger, k.runAsUser, commands, k); err != nil {
+		return err
+	}
+
+	// Start to read from shell and write to datachannel
+	k.logger.Debugf("Start separate go routine to read from pty stdout and write to data channel")
+	done := make(chan int, 1)
+	go func() {
+		done <- k.writePump(k.logger)
+	}()
```

Maybe this is just a placeholder since you haven't fully integrated with the datachannel, so feel free to ignore this comments because you probably already know, but I'm assuming in the future `done` -> `streamOutputChan`?

---

It is just a placeholder in case I want to take some additional action on the write pump ending. 

You are right it is probably better to bubble that up rather than log the error and fail. What is the best path in the code base for saying ``catastrophic error please kill''?

---

We should send a keysplitting error on the message if we fail to start. This will probably need to wait until you integrate with the datachannel tho